### PR TITLE
adding fast and parallel-friendly kdtree implementation

### PR DESCRIPTION
**Pull request by felicepantaleo**
This PR adds a fast/parallel-friendly kd-tree implementation in CommonTools. 
The plan is to replace the existing PF KDtree (based on lists) with this one which is based on vectors and SoA in HGCal wfs, making it friendly for accelerators. 
Searching and building produces exactly the same results, searching with 3D point cloud of about 100k points makes it ~3x faster than the existing implementation.

@rovere @malgeri @VinInn @makortel

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+  8, 12, 20, 28, 15, 17, 24, 7, 19, 27, 23, 6, 26, 5, 4, 31}};
+    unsigned int ilog2 (unsigned int v)
+	    {
+	    v |= v >> 1; // first round down to one less than a power of 2 
+	    v |= v >> 2;
+	    v |= v >> 4;
+	    v |= v >> 8;
+	    v |= v >> 16;
+	    return MultiplyDeBruijnBitPosition[(unsigned int)(v * 0x07C4ACDDU) >> 27];
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: Please improve readability by removing the many empty lines (when not needed to separate methods or logical parts inside the code)

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+	    }
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: Remove till L28

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+{
+
+public:
+
+
+FKDTree()
+	{
+		theNumberOfPoints = 0;
+		theDepth = 0;
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+	}
+
+
+
+	void push_back(const FKDPoint<TYPE, numberOfDimensions>& point)
+	{
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove empty lines here and below the instruction

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+		thePoints.push_back(point);
+
+
+	
+	}
+
+	void push_back(FKDPoint<TYPE, numberOfDimensions> && point)
+	{
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove empty lines here and below the instruction

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+				index = indicesToVisit[visitedindicesThisDepth];
+				intersection = intersects(index, minPoint, maxPoint, dimension);
+				firstSonToVisitNext = leftSonIndex(index);
+            
+
+				if (intersection)
+				{
+					if (is_in_the_box(index, minPoint, maxPoint))
+					{
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+					{
+										
+						foundPoints.emplace_back(theIds[index]);
+					}
+					numberOfSonsToVisitNext = (firstSonToVisitNext < theNumberOfPoints)
+                 		 + ((firstSonToVisitNext + 1) < theNumberOfPoints);
+				}
+				else
+				{
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+
+					firstSonToVisitNext += (theDimensions[dimension][index]
+							< minPoint[dimension]);
+
+              		numberOfSonsToVisitNext = std::min((firstSonToVisitNext< theNumberOfPoints)
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: Indent as in the previous lines

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+			numberOfindicesToVisitThisDepth = indicesToVisit.size();
+			for (unsigned int visitedindicesThisDepth = 0;
+					visitedindicesThisDepth < numberOfindicesToVisitThisDepth;
+					visitedindicesThisDepth++)
+			{
+
+				index = indicesToVisit[visitedindicesThisDepth];
+				intersection = intersects(index, minPoint, maxPoint, dimension);
+				firstSonToVisitNext = leftSonIndex(index);
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+			const FKDPoint<TYPE, numberOfDimensions>& maxPoint,
+			std::vector<unsigned int>& foundPoints)
+	{
+		FQueue<unsigned int> indicesToVisit(16);
+		indicesToVisit.push_back(0);
+		unsigned int index;
+		bool intersection;
+		int dimension;
+		unsigned int numberOfindicesToVisitThisDepth;
+		int numberOfSonsToVisitNext;
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: Why not unsigned int ?

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+
+
+					firstSonToVisitNext += (theDimensions[dimension][index]
+							< minPoint[dimension]);
+
+              		numberOfSonsToVisitNext = std::min((firstSonToVisitNext< theNumberOfPoints)
+                  		+ ((firstSonToVisitNext + 1) < theNumberOfPoints) , 1);
+				}
+
+				for (int whichSon = 0; whichSon < numberOfSonsToVisitNext;
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: both "whichSon" and "numberOfSonsToVisitNext" should better be unsigned int

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+
+
+	void build()
+	{
+
+		theNumberOfPoints =thePoints.size();
+		theDepth = ilog2(theNumberOfPoints);
+		theIntervalLength.resize(theNumberOfPoints, 0);
+		theIntervalMin.resize(theNumberOfPoints, 0);
+      for (unsigned int i = 0; i < numberOfDimensions; ++i)  theDimensions[i].resize(theNumberOfPoints);
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: indent

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+	void search(
+			const FKDPoint<TYPE, numberOfDimensions>& minPoint,
+			const FKDPoint<TYPE, numberOfDimensions>& maxPoint,
+			std::vector<unsigned int>& foundPoints)
+	{
+		FQueue<unsigned int> indicesToVisit(16);
+		indicesToVisit.push_back(0);
+		unsigned int index;
+		bool intersection;
+		int dimension;
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: why not unsigned int ?

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+      theIds.resize(theNumberOfPoints);	
+
+		//gather kdtree building
+		int dimension;
+		theIntervalMin[0] = 0;
+		theIntervalLength[0] = theNumberOfPoints;
+
+		for (unsigned int depth = 0; depth < theDepth; ++depth)
+		{
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 15]
```diff
@@ -0,0 +1,386 @@
...
+		}
+
+		dimension = theDepth % numberOfDimensions;
+		unsigned int firstIndexInDepth = (1 << theDepth) - 1;
+		for (unsigned int indexInArray = firstIndexInDepth;
+				indexInArray < theNumberOfPoints; ++indexInArray)
+		{
+			add_at_position(points[theIntervalMin[indexInArray]],
+					indexInArray);
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+	}
+
+private:
+
+	unsigned int partition_complete_kdtree(unsigned int length)
+	{
+		if (length == 1)
+			return 0;
+		unsigned int index = 1 << (ilog2(length));
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove one line

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+	}
+
+	void add_at_position(FKDPoint<TYPE, numberOfDimensions> && point,
+			const unsigned int position)
+	{
+		for (unsigned int i = 0; i < numberOfDimensions; ++i)
+			theDimensions[i][position] = point[i];
+		theIds[position] = point.getId();
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+		return true;
+	}
+	
+	void add_at_position(const FKDPoint<TYPE, numberOfDimensions>& point,
+			const unsigned int position)
+	{
+		for (unsigned int i = 0; i < numberOfDimensions; ++i)
+			theDimensions[i][position] = point[i];
+		theIds[position] = point.getId();
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: remove

File: CommonTools/RecoAlgos/interface/FQueue.h [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+
+	void clear()
+	{
+
+		theBuffer.clear();
+
+		theSize = 0;
+		theFront = 0;
+		theTail = 0;
+	}
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FQueue.h]: here one empty line could help, instead...

File: CommonTools/RecoAlgos/interface/FQueue.h [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+class FQueue
+{
+public:
+	FQueue()
+	{
+		theSize = 0;
+		theFront = 0;
+		theTail = 0;
+		theCapacity = 0;
+
```
Comment (perrotta) [inline CommonTools/RecoAlgos/interface/FQueue.h]: The same comments about empty lines and indentations also hold for this file (not marked all them)

**Review by perrotta — COMMENTED** [review]
In general, the huge indentation (it is better avoiding computer tabs for it) and the unnecessary and illogical insertion of empty lines and instructions broken in two or more lines, make the review of this code rather painful.
Could you please adjust it, and make it better readable?

**Comment by VinInn** [comment]
Dear @perrotta , I do not disagree on the specific, still we cannot change coding style at each coordinator change.
 Is there a (still valid) cms coding rule about spacing and tabbing?

**Comment by perrotta** [comment]
Ciao @VinInn 
Honestly, I don't know if there there are specific official rules about spacing and tabbing in CMSSW. I only found this piece of code extremely difficult to read and follow on, mostly because of the huge indentations (that force in a few cases to break the same instruction into separate lines) and the wide usage of scattered empty lines. For a possible example of a much better readable code I could cite your own 18408 (just the first example I got at hand)

**Comment by slava77** [comment]
On 4/21/17 3:36 AM, Vincenzo Innocente wrote:
> Dear @perrotta <https://github.com/perrotta> , I do not disagree on the
> specific, still we cannot change coding style at each coordinator change.

I'm curious who was the last coordinator that asked you to code
with 8-char indentation or add extra empty lines inside function scopes.

File: CommonTools/RecoAlgos/interface/FKDTree.h [hunk 21]
```diff
@@ -0,0 +1,386 @@
...
+					indicesToVisit.push_back(firstSonToVisitNext + whichSon);
+			      }
+			}
+
+			indicesToVisit.pop_front(numberOfindicesToVisitThisDepth);
+		}
+	}
+
+
+	void build()
```
Comment (kpedro88) [inline CommonTools/RecoAlgos/interface/FKDTree.h]: The total duplication of the `build()` function just to change the source of the points vector is unnecessary. This function only needs to exist once.

**Comment by felicepantaleo** [comment]
sorry for the late reply, I have implemented all the comments.
Thanks for reviewing it, it was very useful.

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by perrotta** [comment]
@felicepantaleo : since these are new tools, it would be great to have some doxygen-like documentation added to it. 

Moreover, also a unit test should be better provided, just to keep monitored the continued functionality of this code. As such, I can also check that everything works (i.e. it doesn't crashes: honestly, I can hardly dig into the details of your implementation...)

**Comment by VinInn** [comment]
ok, this is really NEW.
never heard about doxigen doc.
WHO is using doxigen this days btw? (I do not even know where to look for it).
I support the unit test

**Comment by davidlange6** [comment]
its existed since ever on autopilot.. though perhaps markdown would be a better choice for documentation these days

http://cmsdoxygen.web.cern.ch/cmsdoxygen/

**Comment by davidlange6** [comment]
I should have removed the "perhaps"...

**Comment by VinInn** [comment]
> On 3 May, 2017, at 6:27 PM, David Lange <notifications@github.com> wrote:
> 
> its existed since ever on autopilot.. though perhaps markdown would be a better choice for documentation these days 
> 
> http://cmsdoxygen.web.cern.ch/cmsdoxygen/ 

it seems to point direclty to git....
cannot find any native doxigen comment...
v.

**Comment by davidlange6** [comment]
I'm sure there is some somewhere. but there isn't much in CMSSW...

**Comment by slava77** [comment]
On 5/3/17 9:34 AM, David Lange wrote:
> I'm sure there is some somewhere. but there isn't much in CMSSW...

I think that even plain C/C++ style comments would already be better
than no interface documentation at all.

**Comment by felicepantaleo** [comment]
ok, I'll add a unit test and a markdown doc file documenting the class

**Comment by perrotta** [comment]
Thank you @VinInn  for providing the unit-test and @felicepantaleo for adding a few lines of documentation.

Code in CommonTools is likely supposed to become a common tool. If you add a few words of explanation/comment to the methods and data members of the interface of your new classes preceeded by (e.g.) "\\\", they can be both accessed in doxygen and also simply read by the poor user who just browses or edits the code, trying to understand how to use it.

**Comment by felicepantaleo** [comment]
@perrotta could you point me to unit tests of some other data structure so that I can imitate?

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
New code made available in CommonTools/RecoAlgos, not accessed by any module in CMSSW by now.
Visual inspection looks ok, and the unit test provided (thank you Felice) ends succesfully.
Full debug and complete check of fuctionality will necessarily only be dealt with once this code will be actually used in future applications.

**Comment by davidlange6** [comment]
+1

**Pull request merged**